libass/ass_parse.c

```diff
@@ -964,59 +964,48 @@ void apply_transition_effects(ASS_Renderer *render_priv, ASS_Event *event)
-    int tm_current;
-    double dt;
-    int x;
-    int x_start, x_end;
+    int tm_current = render_priv->time - render_priv->state.event->Start;
 
-    tm_current = render_priv->time - render_priv->state.event->Start;
-    timing = 0;
-    s1 = s2 = 0;
-    for (i = 0; i <= render_priv->text_info.length; ++i) {
-        cur = render_priv->text_info.glyphs + i;
-        if ((i == render_priv->text_info.length)
-            || (cur->effect_type != EF_NONE)) {
-            s1 = s2;
-            s2 = cur;
-            if (s1) {
-                e1 = s2 - 1;
-                tm_start = timing + s1->effect_skip_timing;
-                tm_end = tm_start + s1->effect_timing;
-                timing = tm_end;
-                x_start = 1000000;
-                x_end = -1000000;
-                for (cur2 = s1; cur2 <= e1; ++cur2) {
-                    x_start = FFMIN(x_start, d6_to_int(cur2->bbox.x_min + cur2->pos.x));
-                    x_end = FFMAX(x_end, d6_to_int(cur2->bbox.x_max + cur2->pos.x));
-                }
+    int timing = 0;
+    GlyphInfo *last_boundary = NULL;
+    for (int i = 0; i <= render_priv->text_info.length; i++) {
+        if (i < render_priv->text_info.length &&
+                render_priv->text_info.glyphs[i].effect_type == EF_NONE)
+            continue;
 
-                dt = (tm_current - tm_start);
-                if ((s1->effect_type == EF_KARAOKE)
-                    || (s1->effect_type == EF_KARAOKE_KO)) {
-                    if (dt >= 0)
-                        x = x_end + 1;
-                    else
-                        x = x_start;
-                } else if (s1->effect_type == EF_KARAOKE_KF) {
-                    dt /= (tm_end - tm_start);
-                    x = x_start + (x_end - x_start) * dt;
-                } else {
-                    ass_msg(render_priv->library, MSGL_ERR,
-                            "Unknown effect type");
-                    continue;
-                }
+        GlyphInfo *start = last_boundary;
+        GlyphInfo *end = render_priv->text_info.glyphs + i;
+        last_boundary = end;
+        if (!start)
+            continue;
 
-                for (cur2 = s1; cur2 <= e1; ++cur2) {
-                    cur2->effect_type = s1->effect_type;
-                    cur2->effect_timing = x - d6_to_int(cur2->pos.x);
-                }
-            }
+        int tm_start = timing + start->effect_skip_timing;
+        int tm_end = tm_start + start->effect_timing;
+        timing = tm_end;
+
+        int x_start = 1000000;
+        int x_end = -1000000;
+        for (GlyphInfo *info = start; info < end; info++) {
+            x_start = FFMIN(x_start, d6_to_int(info->bbox.x_min + info->pos.x));
+            x_end = FFMAX(x_end, d6_to_int(info->bbox.x_max + info->pos.x));
+        }
+
+        int x;
+        if (start->effect_type == EF_KARAOKE ||
+                start->effect_type == EF_KARAOKE_KO) {
```

Comment:
Not sure what's going on with the indent here?

Reply:
It’s a style we use elsewhere (but not everywhere): the second line is indented by two levels. The good thing about it is that for control structures like `if` in particular, it helps separate the condition from the body that follows. The bad thing is that it may look weird at first 😃

I don’t strongly feel either way; I just thought this was the more common style. But I may be wrong about that.

Reply:
After a discussion on IRC, I’ve changed this to line-trailing `||` and a single indent.